File: python/src/semantic_retrieval/document_parsers/multi_document_parser.py
from dataclasses import dataclass
from typing import Optional, Sequence

from semantic_retrieval.common.types import CallbackEvent
from semantic_retrieval.document.document import IngestedDocument, RawDocument
from semantic_retrieval.document.metadata.document_metadata import (
    DocumentMetadata,
)
from semantic_retrieval.document.metadata.document_metadata_db import (
    DocumentMetadataDB,
)
from semantic_retrieval.document_parsers.parser_registry import ParserRegistry
from semantic_retrieval.utils.callbacks import CallbackManager


@dataclass
class ParserConfig:
    metadata_db: Optional[DocumentMetadataDB]
    # access_control_policy_factory: Optional[
    #     DocumentAccessPolicyFactory
    # ]  # TODO [P1]: Add type for access control policy factory in followup diff
    parser_registry: Optional[ParserRegistry] = None


async def parse_documents(
    documents: Sequence[RawDocument],
    metadata_db: Optional[DocumentMetadataDB],
    callback_manager: CallbackManager,
) -> Sequence[IngestedDocument]:
    parser_registry = ParserRegistry()

    ingested_documents = []
    for document in documents:
        parser = parser_registry.get_parser(document.mime_type)
        ingested_document = (await parser.parse(document)).unwrap()

        if metadata_db is not None:
            # access_policies = []
            # if parser_config.access_control_policy_factory:
            #     access_policies = await parser_config.access_control_policy_factory.get_access_policies(
            #         document
            #     )

            await metadata_db.set_metadata(
                document.document_id,
                DocumentMetadata(
                    document_id=document.document_id,
                    uri=document.uri,
                    mime_type=document.mime_type,
                    metadata={},
                    attributes={},
                    # access_policies=access_policies,
                ),
            )

        ingested_documents.append(ingested_document)

    await callback_manager.run_callbacks(
        CallbackEvent(
            name="multi_parse_documents",
            data=dict(ingested_documents=ingested_documents),
        )
    )

    return ingested_documents

File: typescript/src/ingestion/document-parsers/multiDocumentParser.ts
import { DocumentAccessPolicyFactory } from "../../access-control/documentAccessPolicyFactory";
import { IngestedDocument, RawDocument } from "../../document/document";
import { DocumentMetadataDB } from "../../document/metadata/documentMetadataDB";
import { CallbackManager } from "../../utils/callbacks";
import { ParserRegistry } from "./parserRegistry";

type ParserConfig = {
  metadataDB?: DocumentMetadataDB;
  accessControlPolicyFactory?: DocumentAccessPolicyFactory;
  parserRegistry?: ParserRegistry;
  callbackManager?: CallbackManager;
};

/**
 * A basic implementation for parsing Documents from RawDocuments, using
 * a config to determine which specific parsers to use for each MIME type.
 * If a metadataDB and accessControlPolicyFactory are provided, the parsed
 * documents will have relevant metadata and access policies set in the metadataDB.
 * @param rawDocuments An array of RawDocuments to parse into Documents
 * @param config A ParserConfig for configuring the parsing implementation
 * @returns
 */
export async function parseDocuments(
  rawDocuments: RawDocument[],
  config: ParserConfig
): Promise<IngestedDocument[]> {
  const parserRegistry = config.parserRegistry ?? new ParserRegistry(config);

  // Sanity check all parsers are available
  for (const rawDoc of rawDocuments) {
    if (!parserRegistry.getParser(rawDoc.mimeType)) {
      throw new Error("No parser available for MIME type: " + rawDoc.mimeType);
    }
  }

  return await Promise.all(
    rawDocuments.map(async (rawDocument) => {
      const document = await parserRegistry
        .getParser(rawDocument.mimeType)!
        .parse(rawDocument);

      if (config.metadataDB) {
        let accessPolicies;
        if (config.accessControlPolicyFactory) {
          accessPolicies =
            await config.accessControlPolicyFactory.getAccessPolicies(
              rawDocument
            );
        }

        await config.metadataDB.setMetadata(document.documentId, {
          documentId: document.documentId,
          rawDocument,
          document,
          hash: rawDocument.hash,
          uri: rawDocument.uri,
          name: rawDocument.name,
          mimeType: rawDocument.mimeType,
          metadata: {},
          attributes: {},
          accessPolicies,
        });
      }

      return document;
    })
  );
}
